test/unit-test/mqtt_agent_utest.c

```diff
@@ -853,3 +909,194 @@ void test_MQTTAgent_Terminate_success( void )
+
+    /* Making Sure that acknowledgement is added. */
+    TEST_ASSERT_EQUAL( 1, mqttAgentContext.pPendingAcks[ 0 ].packetId );
+    TEST_ASSERT_EQUAL_MEMORY( &command_to_send, mqttAgentContext.pPendingAcks[ 0 ].pOriginalCommand, sizeof( Command_t ) );
+    /* Making Sure that callback is not called. */
+    TEST_ASSERT_EQUAL( 0, commandCompleteCallbackCount );
+}
+
+/**
+ * @brief Test that MQTTAgent_CommandLoop return MQTTNoMemory if pendingAcks array is full.
+ */
+void test_MQTTAgent_CommandLoop_add_acknowledgment_failure( void )
+{
+    MQTTStatus_t mqttStatus;
+    MQTTAgentContext_t mqttAgentContext;
+    MQTTAgentCommandFuncReturns_t pReturnFlags;
+    AckInfo_t * pendingAcks = mqttAgentContext.pPendingAcks;
+    size_t i = 0;
+    Command_t command_to_send;
+
+    setupAgentContext( &mqttAgentContext );
+    mqttAgentContext.mqttContext.connectStatus = MQTTConnected;
+    pReturnFlags.addAcknowledgment = true;
+    pReturnFlags.runProcessLoop = true;
+    pReturnFlags.endLoop = true;
+    pReturnFlags.packetId = 1U;
+
+
+    MQTTAgentCommand_Publish_ExpectAnyArgsAndReturn( MQTTSuccess );
+    MQTTAgentCommand_Publish_ReturnThruPtr_pReturnFlags( &pReturnFlags );
+    MQTT_ProcessLoop_IgnoreAndReturn( MQTTSuccess );
+
+    /* Initializing command to be sent to the commandLoop. */
+    command_to_send.commandType = PUBLISH;
+    command_to_send.pCommandCompleteCallback = stubCompletionCallback;
+    command_to_send.pCmdContext = NULL;
+    command_to_send.pArgs = NULL;
+
+    for( i = 0; i < MQTT_AGENT_MAX_OUTSTANDING_ACKS; i++ )
+    {
+        /* Assigning valid packet ID to all array spaces to make no space for incoming acknowledgment. */
+        pendingAcks[ i ].packetId = i + 1;
+    }
+
+    mqttAgentContext.agentInterface.pMsgCtx->pSentCommand = &command_to_send;
+
+    mqttStatus = MQTTAgent_CommandLoop( &mqttAgentContext );
+
+    TEST_ASSERT_EQUAL( MQTTNoMemory, mqttStatus );
+
+    /* Making Sure that callback is called. */
+    TEST_ASSERT_EQUAL( 1, commandCompleteCallbackCount );
+}
+
+/**
+ * @brief Test mqttEventCallback invocation via MQTT_ProcessLoop.
+ */
+void test_MQTTAgent_CommandLoop_with_eventCallback( void )
+{
+    MQTTStatus_t mqttStatus;
+    MQTTAgentContext_t mqttAgentContext;
+    MQTTAgentCommandFuncReturns_t pReturnFlags;
+    MQTTPacketInfo_t packetInfo;
+    MQTTDeserializedInfo_t deserializedInfo;
+    Command_t command, command_to_send;
+
+    /* Setting up MQTT Agent Context. */
+    setupAgentContext_mqttEventCallback( &mqttAgentContext );
+    mqttAgentContext.mqttContext.connectStatus = MQTTConnected;
+    pReturnFlags.addAcknowledgment = false;
+    pReturnFlags.runProcessLoop = true;
+    pReturnFlags.endLoop = true;
+
+    /* Invoking mqttEventCallback with MQTT_PACKET_TYPE_PUBREL packet type. */
+    packetInfo.type = MQTT_PACKET_TYPE_PUBREL;
+    mqttAgentContext.mqttContext.appCallback( &( mqttAgentContext.mqttContext ), &packetInfo, &deserializedInfo );
```

Comment:
In the long run it might be better to just write a mock for MQTT_ProcessLoop(). That way you can test the API function MQTTAgent_CommandLoop() instead of this function pointer that the user isn't expected to use.